components/compositing/refresh_driver.rs

```diff
@@ -2,129 +2,141 @@
+        }
     }
 
-    /// Notify the [`RefreshDriver`] that a paint is about to happen. This will trigger
-    /// new animation frames for all active `WebView`s and schedule a new frame deadline.
-    pub(crate) fn notify_will_paint(
-        &self,
-        webview_renderers: Values<'_, WebViewId, WebViewRenderer>,
-    ) {
+    pub(crate) fn notify_will_paint(&self, compositor: &mut IOCompositor) {
         // If we are still waiting for the frame to timeout this paint was caused for some
         // non-animation related reason and we should wait until the frame timeout to trigger
         // the next one.
-        if self.waiting_for_frame_timeout.load(Ordering::Relaxed) {
+        if self.waiting_for_frame.load(Ordering::Relaxed) {
             return;
         }
 
-        // If any WebViews are animating ask them to paint again for another animation tick.
-        let animating_webviews: Vec<_> = webview_renderers
-            .filter_map(|webview_renderer| {
-                if webview_renderer.animating() {
-                    Some(webview_renderer.id)
-                } else {
-                    None
-                }
-            })
-            .collect();
+        // Limit the borrow of `self.observers` to the minimum here.
+        let still_has_observers = {
+            let mut observers = self.observers.borrow_mut();
+            observers.retain(|observer| observer.frame_started(compositor));
+            !observers.is_empty()
+        };
 
-        // If nothing is animating any longer, update our state and exit early without requesting
-        // any noew frames nor triggering a new animation deadline.
-        if animating_webviews.is_empty() {
-            self.animating.set(false);
-            return;
+        if still_has_observers {
+            self.observe_next_frame();
         }
+    }
 
-        if let Err(error) =
-            self.constellation_sender
-                .send(EmbedderToConstellationMessage::TickAnimation(
-                    animating_webviews,
-                ))
+    fn observe_next_frame(&self) {
+        self.waiting_for_frame.store(true, Ordering::Relaxed);
+
+        let waiting_for_frame = self.waiting_for_frame.clone();
+        let event_loop_waker = self.event_loop_waker.clone_box();
+        self.refresh_driver.observe_next_frame(Box::new(move || {
+            waiting_for_frame.store(false, Ordering::Relaxed);
+            event_loop_waker.wake();
+        }));
+    }
+
+    /// Whether or not the renderer should trigger a message to the embedder to request a
+    /// repaint. This might be false if: we are animating and the repaint reason is just
+    /// for a new frame. In that case, the renderer should wait until the frame timeout to
+    /// ask the embedder to repaint.
+    pub(crate) fn wait_to_paint(&self, repaint_reason: RepaintReason) -> bool {
+        if self.observers.borrow().is_empty() || repaint_reason != RepaintReason::NewWebRenderFrame
```

Comment:
The condition here doesn't seem to match what is said in the comment. 

> This might be false if: we are animating and the repaint reason is just for a new frame

The code returns false if the reason is anything other than a new WebRender frame.